storio-content-resolver/src/test/java/com/pushtorefresh/storio/contentresolver/impl/RxChangesObserverTest.java

```diff
@@ -0,0 +1,101 @@
+package com.pushtorefresh.storio.contentresolver.impl;
+
+import android.content.ContentResolver;
+import android.database.ContentObserver;
+import android.net.Uri;
+import android.os.Handler;
+
+import com.pushtorefresh.storio.contentresolver.Changes;
+
+import org.junit.Test;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import rx.Observable;
+import rx.Subscription;
+
+import static java.util.Collections.singleton;
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.anyBoolean;
+import static org.mockito.Matchers.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+public class RxChangesObserverTest {
+
+    @Test
+    public void shouldRegisterContentObserverAfterSubscribingToObservable() {
+        ContentResolver contentResolver = mock(ContentResolver.class);
+
+        Observable<Changes> observable = RxChangesObserver
+                .observeChanges(
+                        contentResolver,
+                        singleton(mock(Uri.class)),
+                        mock(Handler.class)
+                );
+
+        // Should not register ContentObserver before subscribing to Observable
+        verify(contentResolver, times(0))
+                .registerContentObserver(any(Uri.class), anyBoolean(), any(ContentObserver.class));
+
+        Subscription subscription = observable.subscribe();
+
+        // Should register Content Observer after subscribing to Observable
+        verify(contentResolver)
+                .registerContentObserver(any(Uri.class), anyBoolean(), any(ContentObserver.class));
+
+        subscription.unsubscribe();
+    }
+
+    @Test
+    public void shouldUnregisterContentObserverAfterUnsubscribingFromObservable() {
+        ContentResolver contentResolver = mock(ContentResolver.class);
+
+        Subscription subscription = RxChangesObserver
+                .observeChanges(
+                        contentResolver,
+                        singleton(mock(Uri.class)),
+                        mock(Handler.class))
+                .subscribe();
+
+        // Should not unregister before unsubscibe from Subscription
+        verify(contentResolver, times(0)).unregisterContentObserver(any(ContentObserver.class));
+
+        subscription.unsubscribe();
+
+        // Should unregister ContentObserver after unsubscribing from Subscription
+        verify(contentResolver).unregisterContentObserver(any(ContentObserver.class));
+    }
+
+    @Test
+    public void shouldRegisterContentObserverForPassedUris() {
+        Set<Uri> uris = new HashSet<Uri>();
+
+        // All Uris are different (objects)
+        uris.add(mock(Uri.class));
+        uris.add(mock(Uri.class));
+        uris.add(mock(Uri.class));
```

Comment:
Equals of mocked objects just compares references, which is okay for such
tests :)
4 июня 2015 г. 3:03 PM пользователь "Dmitrii Nikitin" <
notifications@github.com> написал:

> In
> storio-content-resolver/src/test/java/com/pushtorefresh/storio/contentresolver/impl/RxChangesObserverTest.java
> <https://github.com/pushtorefresh/storio/pull/360#discussion_r31713866>:
>
> > +        verify(contentResolver, times(0)).unregisterContentObserver(any(ContentObserver.class));
> > +
> > +        subscription.unsubscribe();
> > +
> > +        // Should unregister ContentObserver after unsubscribing from Subscription
> > +        verify(contentResolver).unregisterContentObserver(any(ContentObserver.class));
> > +    }
> > +
> > +    @Test
> > +    public void shouldRegisterContentObserverForPassedUris() {
> > +        Set<Uri> uris = new HashSet<Uri>();
> > +
> > +        // All Uris are different (objects)
> > +        uris.add(mock(Uri.class));
> > +        uris.add(mock(Uri.class));
> > +        uris.add(mock(Uri.class));
>
> Yep, I understand, but it wasn't so clearly that you will rely on mocked
> object equality. Also, imho it will be a bit harder to debug.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pushtorefresh/storio/pull/360/files#r31713866>.
>